Split BufferedWedgeInfo into two layers

Denrage opened the pull request:
This PR is a follow up to the problem mentioned here https://github.com/ppy/osu/pull/7540#discussion_r376642332
The main problem is that the BufferedContainer which the BeatmapInfoWedge is using creates the problem that it's not possible to update any Drawables inside it without redrawing the whole wedge.
In this PR i splitted up the Container into a Foreground and Background. The background still uses the BufferedContainer while the foreground can normally be updated.
This ultimately comes with some sort of performance impact because not everything is cached anymore but this would allow us to have a working hover effect on the Author Link (https://github.com/ppy/osu/pull/12488) and also to remove the workaround in https://github.com/ppy/osu/pull/10576, so the StarRatingDisplay can be updated without redrawing the wedge.
Currently i moved everything out except of the actual background even if things don't really change like the VersionLabel.

UI wise nothing should have changed (positioning, transitions, etc.)

One thing that i saw during this PR is that the DifficultyColourBar is not using the same colour as the StarRatingDisplay for ExpertPlus difficulties, but i didn't saw it as a part of this PR

Denrage committed:
splitted updateable part of wedge

Denrage committed:
Added bindable stardifficulty to StarRatingDisplay

Denrage committed:
WIP refresh BPM-Label on mod change

Denrage committed:
removed mods binding passthrough

Denrage committed:
Removed StarDifficulty binding passthrough

Denrage committed:
Fix CodeFactor error

Denrage committed:
Removed unnecessary whitespaces

Denrage committed:
Fixed InspectCode

bdach commented:
> One thing that i saw during this PR is that the DifficultyColourBar is not using the same colour as the StarRatingDisplay for ExpertPlus difficulties

I believe that is partly intentional (helps readability), and also likely to change with future design updates, so I'd say it's unimportant enough to ignore for now.

Denrage commented on osu.Game/Screens/Ranking/Expanded/StarRatingDisplay.cs:
```diff
@@ -22,22 +24,37 @@ namespace osu.Game.Screens.Ranking.Expanded
+        /// </summary>
+        /// <param name="beatmapInfo">The <see cref="BeatmapInfo"/> to use to create a bindable for <see cref="StarDifficulty"/></param>
+        public StarRatingDisplay(BeatmapInfo beatmapInfo)
         {
-            AutoSizeAxes = Axes.Both;
+            this.beatmapInfo = beatmapInfo;
+        }
```
I agree with you in the regard that the StarRatingDisplay shouldn't actually have a dependency on the beatmapInfo but how would you manage updating the star rating? You could Expire the whole container and add a new one with the updated rating but especially in the BeatmapInfoWedge case you would have the problem of sorting the children so the rating is above the StatusPill

bdach replied:
Not sure what the problem is there in particular. Isn't everything recreated on every beatmap change as-is anyway?

The way I see it, the front layer (the not-background one) should either get fully recreated (in which case this ceases to be a problem), or fully update dynamically (so it just gets created once in the start and then only data in it changes), which also makes "children sorting" concerns null and void.

Denrage replied:
The star rating gets also updated on mod change therefor the wedge needs to be updated without changing the beatmap. I tried the fully update dynamically way with the beatmapInfo in the constructor. It could also also have a property with the star rating which can get changed outside and updates itself accordingly then

bdach replied:
Ok, fair point. Would rather have the star rating as a plain public property or bindable on `StarRatingDisplay`, then.

Denrage commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -515,41 +542,58 @@ public InfoLabel(BeatmapStatistic statistic)
-                {
-                    this.difficulty = difficulty;
-                }
+        public class BufferedWedgeBackground : BufferedContainer
```
So you would generalize it as a whole and not specific to the BeatmapInfo?

bdach replied:
Again not sure what you mean by that. In here I'm asking mostly for mechanical changes (split to own file, make `internal`, nest the `BufferedContainer` in a `CompositeDrawable` rather than using straight inheritance). No idea what "generalisation" would be in this context.

Denrage replied:
To have a wedge background also for other components to use it but yeah we can let it tailored to the beatmap only

Denrage commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -127,6 +111,10 @@ void removeOldInfo()
                     Info?.FadeOut(250);
                     Info?.Expire();
                     Info = null;
+
+                    Background?.FadeOut(250);
+                    Background?.Expire();
+                    Background = null;
```
Sure, can do. I can't grasp the overhead of using a "whole" container "just" to reduce code duplication

bdach replied:
What I mean is to have, using a tree structure:

```
Container
	Info
	Background
```

Then you just fade/expire the uppermost `Container` rather than both pieces individually.

Denrage replied:
Yeah understood that, i only don't know how much overhead a container produces. If it's completely negligible, ofc it's better structurewise, if it has an impact i dont know how much it is and when it justifies it to reduce code duplication, i just cant measure it, therefor i will completely trust you on that and will do it

peppy commented:
Not checking the code just yet as there are pending refactors, but visually this has slightly regressed the transition. At a glance it looks like the new and old text layer is at the same depth, which makes it *above* the background layer, causing it to look busier than it should.

Correct order should be:

```csharp
old background
old text

new background
new text
```

but I believe it currently is


```csharp
old background
new background

old text
new text
```

Denrage commented:
@peppy should be fixed with putting both layers into one container and transition the whole container instead the layers individually (which is part of the refactor)

Denrage committed:
Fixed race condition in StarRatingDisplay

Denrage committed:
Moved Info and Background into own container

Denrage committed:
Removed BeatmapInfo in StarRatingDisplay

Denrage committed:
Refactored background of BeatmapInfoWedge

Denrage committed:
Codestyle fixes

bdach commented on osu.Game/Screens/Ranking/Expanded/StarRatingDisplay.cs:
```diff
@@ -47,9 +61,20 @@ private void load(OsuColour colours, BeatmapDifficultyCache difficultyCache)
                 ? ColourInfo.GradientVertical(Color4Extensions.FromHex("#C1C1C1"), Color4Extensions.FromHex("#595959"))
                 : (ColourInfo)colours.ForDifficultyRating(difficulty.DifficultyRating);
 
+            colorContainer.Colour = backgroundColour;
+
+            wholePartText.Text = $"{wholePart}";
+            fractionPartText.Text = $"{separator}{fractionPart}";
```
These changes have removed some fade transitions that were naturally there due to how the component was originally structured (everything in a buffered container), and therefore this looks worse now. To understand what I mean try toggling mods on `master` and this branch (so that the colour of the star rating changes).

Not sure whether it's worth holding this pull up for that as I understand a redesign of the song select screen is in the works, but yeah.

Denrage replied:
Hmm, that's true. Things that come to my mind to restore the previous behavior on rating change:
- Fade out sprite texts and set alpha back to 100% (don't know if it's a thing to do smth like that)
- Fade out whole text container and set alpha back
    - Both would have the problem that the fading is set inside `StarRatingDisplay` and therefore not possible to set it from outside
-  Fade out the top right metadata container (Containing `StarRatingDisplay` and `StatusPill`) in `BeatmapWedgeInfo` and recreate it
- Fade out and recreate the whole Info container

or your last point `ignore it for now`. I don't know what the best way is to handle this case

bdach replied:
Would probably be best to incorporate the transition into the component itself:

* Background would simply use `.FadeColour()`
* As for the foreground, you'd probably need to `.FadeOut().Expire()` the old sprite text and create a new one with `.FadeIn()` to preserve the visual effect.

Might be a bit finicky in places because transforms cannot be added in `load()`, so you'd need some `IsLoaded` checks. It's a pretty rudimentary pattern though, should find many usages around.

Denrage replied:
`.FadeOut(250).Expire()` results in a weird transition where the size of the container gets larger and smaller:
https://streamable.com/1cgb0z
I implemented it now w/o the expire and it feels somewhat good. Here is a comparison with the transition on master:
https://streamable.com/3o1nv1

bdach replied:
The expire was supposed to be an integral part of this. It was to make sure the old texts are cleaned up after the fades have completed. Have you checked whether they really are, using the draw visualiser?

Denrage replied:
ohh, i misswrote i meant i did it now without the fadeout the expire is still there and used

bdach commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -135,97 +118,96 @@ void removeOldInfo()
         }
 
-        protected override void Dispose(bool isDisposing)
+        public class BeatmapInfoWedgeContainer : Container
```
The ordering changes in this file are making this diff unnecessarily large and making it harder to review the pull (have to cross-check implementations for changes). Would rather see member order restored to the original where it's been altered to decrease diff size.

Denrage committed:
Reorganize methods

Denrage committed:
Added transition in StarRatingDisplay

peppy committed:
Merge branch 'master' into split-BufferedWedgeInfo

peppy commented on osu.Game/Screens/Ranking/Expanded/StarRatingDisplay.cs:
```diff
@@ -22,86 +22,117 @@ namespace osu.Game.Screens.Ranking.Expanded
     public class StarRatingDisplay : CompositeDrawable
     {
-        private readonly StarDifficulty difficulty;
+        [Resolved]
```
What's the reasoning for any change in this file? Please revert all changes unless there's valid reason (I cannot see any).

Denrage replied:
The reasoning behind it is discussed in this conversation: https://github.com/ppy/osu/pull/12503#discussion_r618753148
The `StarRatingDisplay` was recreated as a whole before because the whole wedge was redrawn, it still needs to be updated and have a transition and therefore it was discussed how we want to accomplish that

peppy replied:
It still is recreated every beatmap change in your code.

Denrage replied:
Yes but it also needs to change when the user changes the mod (Double-Time f.e.) to adjust to the new difficulty it creates. Before the whole wedge got recreated on a mod changes this does not happen anymore

peppy replied:
Want to split that out into its own change? Judging by the issues I found with it, there's too much being changed in this PR.

peppy commented:
The transition is still wrong.


https://user-images.githubusercontent.com/191335/116961996-cbb12200-acdf-11eb-8ce9-5c7cd06043bc.mp4


This should not happen.

peppy commented:
Can crash due to a nullref in `StarDifficultyDisplay`. Likely not an issue if you revert your changes though:


https://user-images.githubusercontent.com/191335/116962137-2c405f00-ace0-11eb-9748-354239ebbcd9.mp4

```csharp
[runtime] 2021-05-04 04:53:19 [debug]: ClickEvent(Left) handled by select no beatmaps available! beatmap.
[runtime] 2021-05-04 04:53:19 [debug]: MouseClick handled by select no beatmaps available! beatmap.
[runtime] 2021-05-04 04:53:19 [debug]: MouseDownEvent(Left) handled by BasicScrollContainer.
Console output is being limited. Please check runtime.log for full logs.
[runtime] 2021-05-04 04:53:20 [error]: An unhandled error has occurred.
[runtime] 2021-05-04 04:53:20 [error]: System.NullReferenceException: Object reference not set to an instance of an object.
[runtime] 2021-05-04 04:53:20 [error]: at osu.Game.Screens.Ranking.Expanded.StarRatingDisplay.setDifficulty(StarDifficulty difficulty) in /Users/dean/Projects/osu/osu.Game/Screens/Ranking/Expanded/StarRatingDisplay.cs:line 55
[runtime] 2021-05-04 04:53:20 [error]: at osu.Game.Screens.Ranking.Expanded.StarRatingDisplay.set_StarDifficulty(StarDifficulty value) in /Users/dean/Projects/osu/osu.Game/Screens/Ranking/Expanded/StarRatingDisplay.cs:line 38
[runtime] 2021-05-04 04:53:20 [error]: at osu.Game.Screens.Select.BeatmapInfoWedge.WedgeInfoText.updateStarRatingDisplay(ValueChangedEvent`1 valueChanged) in /Users/dean/Projects/osu/osu.Game/Screens/Select/BeatmapInfoWedge.cs:line 323
[runtime] 2021-05-04 04:53:20 [error]: at osu.Framework.Bindables.Bindable`1.TriggerValueChange(T previousValue, Bindable`1 source, Boolean propagateToBindings, Boolean bypassChecks)
[runtime] 2021-05-04 04:53:20 [error]: at osu.Framework.Bindables.Bindable`1.set_Value(T value)
[runtime] 2021-05-04 04:53:20 [error]: at osu.Game.Beatmaps.BeatmapDifficultyCache.<>c__DisplayClass26_1.<updateBindable>b__1() in /Users/dean/Projects/osu/osu.Game/Beatmaps/BeatmapDifficultyCache.cs:line 235
[runtime] 2021-05-04 04:53:20 [error]: at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
[runtime] 2021-05-04 04:53:20 [error]: at osu.Framework.Threading.Scheduler.Update()

```

Denrage commented:
> The transition is still wrong.
> 
>  20210504.135100.dotnet.mp4 
> This should not happen.

This happens primarly because the beatmap gets recreated in that test and the check `if (beatmap == value)` in the setter of `public WorkingBeatmap Beatmap` is returning false. I can add that transitions inside the container (in `DifficultyColourBar` f.e.) will only happen if the actual value on which they transition also changes, but what happens in the test shouldn't happen in the game, or rather i cant think of a way how that could happen in normal circumstances

peppy commented:
> but what happens in the test shouldn't happen in the game, or rather i cant think of a way how that could happen in normal circumstances

It happens on every beatmap switch.

peppy committed:
Remove unnecessary pixelSnapping parameter

This was only required because there was text being rendered to the
`BufferedContainer` content until now. Removing this should allow for
better resolution in the background display (due to using a better
minify scale mode).

Denrage committed:
Refactored out changes in StarRatingDisplay

peppy committed:
Remove unnecessary redirection property to `Container.Info`

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -517,38 +516,64 @@ public InfoLabel(BeatmapStatistic statistic)
             private class DifficultyColourBar : Container
             {
-                private readonly StarDifficulty difficulty;
+                [Resolved]
```
This should also be reverted. It has issues and I'd rather address them in a separate PR.

Denrage committed:
Refactored out changes in DifficultyColourBar

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -319,28 +284,80 @@ private void load(LocalisationManager localisation)
+            private DifficultyColourBar createDifficultyColourBar(StarDifficulty difficulty)
+                => new DifficultyColourBar(difficulty)
+                {
+                    RelativeSizeAxes = Axes.Y,
+                    Width = 20,
+                };
+
+            private FillFlowContainer createTopRightMetadataContainer(BeatmapInfo beatmapInfo, StarDifficulty difficulty)
+            {
+                var container = new FillFlowContainer
+                {
+                    Direction = FillDirection.Vertical,
+                    AutoSizeAxes = Axes.Both,
+                };
+
+                if (difficulty.Stars > 0)
                 {
-                    Margin = new MarginPadding { Bottom = 5 }
+                    container.Add(new StarRatingDisplay(difficulty)
+                    {
+                        Anchor = Anchor.TopRight,
+                        Origin = Anchor.TopRight,
+                        Shear = -wedged_container_shear,
+                        Margin = new MarginPadding { Bottom = 5 }
+                    });
                 }
-                : Empty();
+
```
can all this be restored? i can't read this.

peppy replied:
also we really don't need private functions to create single drawables.

Denrage replied:
What do you want to have restored? The `StarRatingDisplay` itself is the same as before, it just needs to get recreated on a difficulty change, the same for `DifficultyColourBar` otherwise both won't change if you change mods. Before the whole wedge was recreated on a mod change. I could recreate the whole info layer on a mod change if you would prefer that.
I recreate the whole container which contains the `StarRatingDisplay` and the `StatusPill` to maintain the order. I could also wrap a container around the `StarRatingDisplay` and expire only the Child of that

I added a private function bc i create the `DifficultyColourBar` at different places. Maybe i can skip it in the `load` function and do a check if a child exists before `Expire`

peppy replied:
You're adding more complexity around the place. Everything should pass `StarDifficulty` as it was previously, not the `beatmapInfo`.

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -319,28 +284,80 @@ private void load(LocalisationManager localisation)
+            }
 
-                addInfoLabels();
+            private void updateDifficulty(ValueChangedEvent<StarDifficulty?> valueChanged)
```
i'd hope this is restored too

Denrage replied:
Yes `addInfoLabels` is further above (https://github.com/ppy/osu/pull/12503/files#diff-10508e952755970ee0b1dfc1cfaac15497e2e8153598c73c0133e3a3e367ca5fR287)

peppy replied:
I mean removing all this star difficulty update logic. It should be in a separate PR.

This PR should *only do what it says in the title*. Try reading the diff yourself. It's legit impossible to parse right now.

Denrage replied:
Yes but the wedge shouldn't loose functionality. I somewhat need to recreate the StarDifficulty and the bar otherwise they won't be updatet anymore on a mod change. On the other hand i also can recreate the whole info layer on a mod change if you would prefer that for this pr

Denrage committed:
reduced complexity

Denrage committed:
Reorganized elements for readability

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -99,20 +88,13 @@ public WorkingBeatmap Beatmap
                 if (beatmap == value) return;
 
                 beatmap = value;
-                cancellationSource?.Cancel();
-                cancellationSource = new CancellationTokenSource();
-
-                beatmapDifficulty?.UnbindAll();
-                beatmapDifficulty = difficultyCache.GetBindableDifficulty(beatmap.BeatmapInfo, cancellationSource.Token);
-                beatmapDifficulty.BindValueChanged(_ => updateDisplay());
```
If you can bring back this logic and make things work (hopefully reducing the lines change in this PR by 50% in the process) that would be a good starting point.

Denrage committed:
Reverted difficulty and mod updates

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -552,5 +520,35 @@ private void load(OsuColour colours)
             }
         }
+
+        public class BeatmapInfoWedgeContainer : Container
```
This naming is confusing.

You currently have a `BeatmapInfoWedge` creating a `BeatmapInfoWedgeContainer` which has `WedgeInfoText`. The ordering of these classes is 1-3-2. One doesn't contain the prefix, while the other two do. The "Container" is inside the non-container (which is back to front to how we usually have it) and I don't think the term container suits in this use case.

Denrage replied:
On second glance the `BeatmapInfoWedgeContainer` isn't actually needed. It is not really doing anything to justify to be it's own class. Would be enough to change https://github.com/ppy/osu/pull/12503/files#diff-10508e952755970ee0b1dfc1cfaac15497e2e8153598c73c0133e3a3e367ca5fR137 to a normal `Container` instantiation, right?

peppy replied:
I think that will help a lot, yes.

Denrage committed:
Removed unnecessary class for wrapping

peppy committed:
Rename Container to DisplayedContent

peppy committed:
Make setters private for protected containers

peppy commented:
Made a few minor amendments, but looks good otherwise. Let me know if you have any questions about the path forward from here (I think you likely already have a good idea). Just try to keep each component update in its own PR (separate PRs for `StarRatingDisplay` and the colour bar, where they both become pieces of the wedge that are not reconstructed each beatmap change).

Note that the fading issues I touched on are even visible with the main text content in the case where it matches the previous content (becomes most obvious). So we're going to need to look into a solution for that as well. It may be that all these pieces are best handled in one subcomponent, or separate, but it's hard to say without trying it out.
